### ISSUE TITLE
`fn` nested in `const fn` causes "Cannot create local mono-item for ..." ICE cross-crate.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#53451]
```rust
// a.rs
#![feature(const_fn)]

pub const fn foo() -> fn() {
    fn bar() {}
    bar
}
```
```rust
// b.rs
extern crate a;
fn main() {
    a::foo()();
}
```
```
error: internal compiler error: librustc_mir/monomorphize/collector.rs:757:
Cannot create local mono-item for DefId(10/0:4 ~ a[8787]::foo[0]::bar[0])
```

Adding `#[inline]` to `bar`'s definition works around the ICE.
Also, the bug doesn't trigger if `foo` is an `#[inline]` non-`const` `fn`, so it's not just that `foo` is inlined cross-crate, but `const fn`-specific behavior is also involved.

cc @oli-obk @michaelwoerister

**Comment by eddyb** [comment]
Still happens on the latest nightly. cc @rust-lang/compiler 
It doesn't block #49219, but the workaround in that PR is a bit ugly.

**Comment by oli-obk** [comment]
I have a fix in https://github.com/rust-lang/rust/pull/56024
